UniversalAlgebra/Complexity/FiniteCSP.lagda

```diff
@@ -0,0 +1,138 @@
+numbers for variable symbols, so
+
+V : Vec ℕ n
+V = [0 1 2 ... n-1]
+
+We can use the following range function to construct the vector of variables.
+
+\begin{code}
+
+range : (n : ℕ) → Vec ℕ n
+range n = tabulate toℕ
+-- `range n` is 0 ∷ 1 ∷ 2 ∷ … ∷ n-1 ∷ []
+
+\end{code}
+
+Let `nvar` denote the number of variables.
+
+A *constraint* consists of
+
+1. a natural number `∣s∣` denoting the number of variables in the "scope" of the constraint;
+
+2. a scope vector,  `s : Vec (Fin nvar) ∣s∣` , where `s i` is `k` if `k` is the i-th variable
+   in the scope of the constraint.
+
+3. a constraint relation, rel, which is a collection of functions mapping (indices of)
+   variables in the scope to elements of the domain.
+
+To summarize, a constraint is a triple (∣s∣ , s , rel), where
+* ∣s∣ is the number of variables in scope s.
+* s is the scope function: s i ≡ v iff v is the i-th variable in scope s.
+* rel is the contraint relation: a collection of functions mapping indices
+  of scope variables to elements in the domain.
+
+\begin{code}
+
+open DecSetoid
+open Fin renaming (zero to fzer ; suc to fsuc)
+
+record Constraint (nvar : ℕ) (dom : Vec (DecSetoid α ℓ) nvar) : Type α where
+ field
+  ∣s∣ : Fin nvar                     -- The "number" of variables involved in the constraint.
+  s : Vec (Fin nvar) (toℕ ∣s∣)        -- Vec of variables involved in the constraint.
+  rel : ((i : Fin (toℕ ∣s∣)) → Carrier ((lookup dom) (lookup s i))) →  Bool
+         -- `rel f` returns true iff the function f belongs to the relation
+
+ satisfies : (∀ i → Carrier ((lookup dom) i)) → Bool         -- An assignment f of values to variables
+ satisfies f = rel (λ (i : Fin (toℕ ∣s∣)) → f (lookup s i))  -- *satisfies* the constraint provided
+                                                             -- the function f, evaluated at each variable
+                                                             -- in the scope, belongs to the relation rel.
+
+
+-- utility functions --
+foldleft : ∀ {α β} {A : Type α} {B : Type β} {m} →
+           (B → A → B) → B → Vec A m → B
+foldleft _⊕_ b []       = b
+foldleft _⊕_ b (x ∷ xs) = foldleft _⊕_ (b ⊕ x) xs
+-- cf. stdlib's foldl, which seems harder to use than this one
+
+bool2nat : Bool → ℕ
+bool2nat false = 0
+bool2nat true = 1
+
+-- The number of elements of v that satisfy P
+countBool : {n : ℕ}{A : Set α} → Vec A n → (P : A → Bool) → ℕ
+countBool v P = foldleft _+_ 0 (map (bool2nat ∘ P) v)
+-- cf. stdlib's count, which works with general predicates (of type Pred A _)
+
+-- Return true iff all elements of v satisfy predicate P
+AllBool : ∀{n}{A : Set α} → Vec A n → (P : A → Bool) → Bool
+AllBool v P = foldleft _∧_ true (map P v)
+-- cf. stdlib's All, which works with general predicates (of type Pred A _)
+
+
+
+open Constraint
+record CSPInstance (nvar : ℕ) (dom : Vec (DecSetoid α ℓ) nvar) : Type α where
+ field
+  ncon : ℕ      -- the number of constraints involved
+  constr : Vec (Constraint nvar dom) ncon
```

Comment:
So I would actually go with a list of constraints, until knowing the number of constraints involved in advance actually makes a difference. My guess is, it won't matter.

Reply:
So, List is better than Vec here?  Any particular reason why?

Reply:
`Vec` is useful when knowing the length 'statically' matters - otherwise might as well go with `List`.

Reply:
Ooops, I forgot to convert the vector of constraints to a list of constraints!  I'll do that now.